tv_apps/smart-system/js/sync_state_machine.js

```diff
@@ -0,0 +1,151 @@
+    enable: {
+      'disabled': 'enabling',
+      'errored': 'enabled'
+    },
+    disable: {
+      'enabled': 'disabled',
+      'errored': 'disabled',
+      'syncing': 'disabled'
+    },
+    sync: {'enabled': 'syncing'},
+    success: {
+      'enabling': 'enabled',
+      'syncing' : 'enabled'
+    },
+    error: {
+      'enabled' : 'errored',
+      'enabling': 'errored',
+      'syncing' : 'errored'
+    }
+  };
+
+  /**
+   * Creates the state transition function per each event.
+   *
+   * For each state transition we trigger an event with name 'onSTATE' where
+   * STATE is the name of the state where the machine transitioned to.
+   * This event gets the name of the transition and the previous state as
+   * arguments along with the additional arguments specific to the transition:
+   *
+   * {
+   *   transition: <string>, // The name of the action that caused the
+   *                         // transition.
+   *   from: <string>  // The previous state the machine was before this
+   *                   // transition.
+   * }
+   *
+   * Example of an action 'A' causing a transition from state 'X' to state 'Y':
+   *
+   * SyncStateMachine.A('dummyArg');
+   *
+   * SyncStateMachine.onY = function(event, from) {
+   *  alert(event); // 'A'
+   *  alert(from); // 'X'
+   *  alert(arguments); // '{'0': 'A', '1': 'X', '2': 'dummyArgs'}'
+   * };
+   *
+   * We will end up with a SyncStateMachine object like:
+   * {
+   *    // Transitions.
+   *    enable: function() {},
+   *    disable: function() {},
+   *    sync: function() {},
+   *    error: function() {},
+   *    success: function() {},
+   *
+   *    // Event handlers.
+   *    onenabled: function(event, from) {},
+   *    ondisabled: function(event, from) {},
+   *    onenabling: function(event, from) {},
+   *    onsyncing: function(event, from) {},
+   *    onerrored: function(event, from) {}
+   * }*
+   */
+  function createTransition(name, transition) {
+    return (...args) => {
+      // Checks if the transition is valid in the current state.
+      if (!transitions[name][SyncStateMachine.state]) {
+        throw new Error('Transition ' + name +
+                        ' invalid for the current state');
+      }
+
+      var from = SyncStateMachine.state;
+      var to = transition[from];
+
+      if (from == to) {
+        return;
+      }
+
+      // Transition to the next state.
+      SyncStateMachine.state = to;
+
+      // Publish the onSTATE event.
+      var onstate = SyncStateMachine['on' + SyncStateMachine.state];
+      if (!onstate || !(onstate instanceof Function)) {
+        return;
+      }
+      onstate.apply(SyncStateMachine, [from, name].concat(args));
+    };
+  }
+
+  for (var name in transitions) {
+    if (!transitions.hasOwnProperty(name)) {
+      continue;
+    }
+    SyncStateMachine[name] = createTransition(name, transitions[name]);
+  }
+
+  // Exported only for testing purposes.
+  exports.SyncStateMachine = SyncStateMachine;
```

Comment:
Guess this line should be added above the comment since it's needed in SyncManager?